addons/mail/models/mail_canned_response.py

```diff
@@ -0,0 +1,88 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models, api
+
+
+class MailCannedResponse(models.Model):
+    """
+    Canned Response: content that will automatically replace the shortcut of your choosing. This content can still be adapted before sending your message.
+    """
+
+    _name = "mail.canned.response"
+    _description = "Canned Response"
+    _order = "id desc"
+    _rec_name = "source"
+
+    source = fields.Char(
+        "Shortcut", required=True, index="trigram",
+        help="Canned response that will automatically be substituted with longer content in your messages."
+        " Type ':' followed by the name of your shortcut (e.g. :hello) to use in your messages.",
+    )
+    substitution = fields.Text(
+        "Substitution",
+        required=True,
+        help="Content that will automatically replace the shortcut of your choosing. This content can still be adapted before sending your message.",
+    )
+    description = fields.Char("Description")
+    last_used = fields.Datetime("Last Used", help="Last time this canned_response was used")
+    group_ids = fields.Many2many("res.groups", string="Authorized Groups", groups="mail.group_mail_canned_response_admin")
+    is_shared = fields.Boolean(
+        string="Determines if the canned_response is currently shared with other users",
+        compute="_compute_is_shared",
+        store=True,
+    )
+    is_editable = fields.Boolean(
+        string="Determines if the canned response can be edited by the current user",
+        compute="_compute_is_editable"
+    )
+
+    @api.depends("group_ids")
+    def _compute_is_shared(self):
+        for canned_response in self:
+            # sudo: regular user should be able to compute
+            canned_response.is_shared = bool(canned_response.sudo().group_ids)
+
+    @api.depends_context('uid')
+    @api.depends("create_uid")
+    def _compute_is_editable(self):
+        for canned_response in self:
+            if not canned_response.id:
+                canned_response.is_editable = True
+            else:
+                try:
+                    canned_response.check_access_rights("write")
+                    canned_response.check_access_rule("write")
```

Comment:
If I filter the canned response in the `for` loop I can't write `False` on the canned response that failed the access rule.
If I set `is_editable` to `False` by default I can't create a new canned response since the editable is always `False`.

This filter is good to know.

Reply:
I don't really like this. Everything is more complicated and obscure, you can't just read it, and understand it, you have to process it to understand what it's doing. That is going to make debugging harder.
That being said, it's looking more efficient?  I suspect that the for loop is just hidden now.
I don't know, I ll trust you on that.

But it's not working for the creation of new records. I don't feel like having `True` as the default value of this compute.

Reply:
That's standard python code actually. There is nothing obscure playing with recordsets.

Reply:
Congratulations, you make me hate Python more! 

It would never occur to me that doing `my_var.my_property = "string"`  could even mean assigning "string" to my_property  for every object inside the `my_var` array.

Reply:
Indeed recordsets are not dictionaries. As we are an ERP (yes), we generally want to do things in batch, like batch updating, hence we can effectively update records in batch. All python code should be thought as being performant in batch.

``Congratulations, you make me hate Python more!``

Well consider you learned how to use recordsets and the basics of ACLs today (hopefully). You should be happy to have learned things ! And that way we won't have to fight that much on next reviews.